core/trino-spi/src/main/java/io/trino/spi/connector/JoinApplicationResult.java

```diff
@@ -13,21 +13,35 @@
  */
 package io.trino.spi.connector;
 
+import io.trino.spi.expression.Variable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Collections.unmodifiableMap;
 import static java.util.Objects.requireNonNull;
 
-public final class JoinApplicationResult
+public final class JoinApplicationResult<T>
 {
-    private final ConnectorTableHandle tableHandle;
+    private final T handle;
+    private final Map<Variable, ColumnHandle> assignments;
```

Comment:
Variable does not help here. It could also be String. The idea is that connector returns new assignments which should contain all the keys from both `leftAssignments` and r`ightAssignments` from the call to `applyJoin`, but mapping those to new column handles.

Why would we need two maps? Can we have duplicate keys?

Reply:
My idea about leftAssignments. rightAssignments was to resolve variables in join conditions.
So if i have one join condition, those maps could be 1-element each.

> Why would we need two maps? Can we have duplicate keys?

when joining between `orders` and `orders`, the column handles can easily be equal between the tables, but we need to keep them distinguished.

Reply:
After discussing offline I went with `ColumnHandle` -> `ColumnHandle` mapping in JoinApplicationResult.java.